crates/ruff_linter/src/rules/ssort/dependencies.rs

```diff
@@ -0,0 +1,785 @@
+use std::fmt;
+
+use rustc_hash::FxHashSet;
+
+use ruff_python_ast::name::Name;
+use ruff_python_ast::visitor::{Visitor, walk_expr};
+use ruff_python_ast::{Expr, Stmt};
+
+/// An error indicating that a cycle was detected in the dependency graph.
+#[derive(Debug)]
+pub(super) struct CycleError {
+    /// The indices of nodes involved in the cycle, in call order.
+    pub(crate) cycle: Vec<usize>,
+}
+
+/// Allows CycleError to be treated as a standard error.
+impl std::error::Error for CycleError {}
+
+/// Allows CycleError to be formatted for display.
+impl fmt::Display for CycleError {
+    /// Format the cycle error for display.
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(formatter, "Cycle detected involving nodes {:?}", self.cycle)
+    }
+}
+
+/// A statement in the dependency graph.
+#[derive(Debug)]
+pub(super) struct Node<'a> {
+    /// The index of the statement in the original statement list.
+    pub(crate) index: usize,
+
+    /// The name of the statement.
+    pub(crate) name: Option<&'a Name>,
+
+    /// The statement in the AST.
+    pub(crate) stmt: &'a Stmt,
+
+    /// Whether this node can be moved during sorting.
+    pub(crate) movable: bool,
+}
+
+/// A visitor that collects all variable names referenced in expressions.
+#[derive(Debug)]
+struct DependencyVisitor<'a> {
+    /// The set of variable names referenced.
+    names: FxHashSet<&'a Name>,
+}
+
+/// Allows DependencyVisitor to be used as a visitor for AST nodes.
+impl<'a> Visitor<'a> for DependencyVisitor<'a> {
+    /// Visit an expression node during AST traversal.
+    ///
+    /// ## Arguments
+    /// * `expr` - The expression node to visit.
+    fn visit_expr(&mut self, expr: &'a Expr) {
```

Comment:
Does this visitor correctly handle scopes? What if two functions define the same variables but they're nested within each other? What about type vars?

Reply:
From my understanding, `ssort` is intentionally scope-agnostic as it only tracks text references, so as long as the text reference matches another function, this should work.

If two functions define the same variables but they're nested then this shouldn't handle they would be considered a separate suite (although I'd argue that is a problem with your code as you shouldn't name two things with the same name).

I'm not sure if type variables will work as I don't think `visit_expr()` handles annotations, but I could be wrong